R: support structs

james-atkins opened the pull request:
This PR allows structs to be returned when running queries from R. Resolves #3334.

C++ side is pretty easy: I recursively build an R list by deferring to the allocate/transform methods of a struct column's children.

~~The R side is trickier... I convert a duckdb struct column to an R data frame which I think is the most intuitive thing to do. However base R data frames really struggle with nested data whereas tibbles support it well. So I've added an option to dbConnect (defaults to false) allowing output to be returned as tibbles rather than data frames. (Personally I favour tibbles everywhere but this [upsets](https://github.com/r-dbi/odbc/issues/3) some people and is an extra dependency.)~~

The R output is now a just a plain old data frame, not a tibble. If the C++ code sets the necessary attributes, then nested and packed columns seem to be supported okay in base R (but for some reason they are difficult to *construct* in base R; `as.data.frame` in particular chokes).

james-atkins commented:
Downside of an opt-in tibble flag is that if someone runs a query from R returning structs there is a nasty warning/error message from data.frame - should I add a warning?

james-atkins commented:
Examples of what I mean by data.frame struggles with nested data:

``` r
library(DBI)
con_df <- dbConnect(duckdb::duckdb(), tibble = FALSE)
con_tibble <- dbConnect(duckdb::duckdb(), tibble = TRUE)

# data.frame promotes struct types to main columns (incorrectly, in my opinion). Tibble does not.
dbGetQuery(con_df, "SELECT 1, {'x': 100, 'y': 'a'}")
#>   1 main.struct_pack(100, 'a').x main.struct_pack(100, 'a').y
#> 1 1                          100                            a
dbGetQuery(con_tibble, "SELECT 1, {'x': 100, 'y': 'a'}")
#> # A tibble: 1 × 2
#>     `1` `main.struct_pack(100, 'a')`$x $y   
#>   <int>                          <int> <chr>
#> 1     1                            100 a

# data.frame cannot handle structs by themselves
dbGetQuery(con_df, "SELECT {'x': 1, 'y': 'a'} AS lone_struct")
#> Warning in format.data.frame(if (omit) x[seq_len(n0), , drop = FALSE] else x, :
#> corrupt data frame: columns will be truncated or padded with NAs
#> Error in match.names(clabs, names(xi)): names do not match previous names
dbGetQuery(con_tibble, "SELECT {'x': 1, 'y': 'a'} AS lone_struct")
#> # A tibble: 1 × 1
#>   lone_struct$x $y   
#>           <int> <chr>
#> 1             1 a

# data.frame cannot handle lists of structs
dbGetQuery(con_df, "SELECT 1, [{'x': 100}] AS s")
#>   1   s
#> 1 1 100
dbGetQuery(con_tibble, "SELECT 1, [{'x': 100}] AS s")
#> # A tibble: 1 × 2
#>     `1` s               
#>   <int> <list>          
#> 1     1 <tibble [1 × 1]>
```

krlmlr commented on tools/rpkg/R/Result.R:
```diff
@@ -67,6 +71,24 @@ list_to_df <- function(x) {
 }
 
+list_to_tibble <- function(x) {
+  if (!requireNamespace("tibble", quietly = TRUE)) {
```
Should the availability of the tibble package be checked when establishing the connection?

krlmlr commented on tools/rpkg/R/Result.R:
```diff
@@ -67,6 +71,24 @@ list_to_df <- function(x) {
+  }
+
+  # Convert any sub data frames to tibbles too
+  sub_tibbled <- recursive_as_tibble(x)
```
I wonder if this can be handled by the C++ layer instead?

krlmlr commented on tools/rpkg/src/statement.cpp:
```diff
@@ -202,6 +205,24 @@ static SEXP allocate(const LogicalType &type, RProtector &r_varvalue, idx_t nrow
+			dest_list.push_back(cpp11::named_arg(name.c_str()) = std::move(dest_child));
+		}
+
+		dest_list.attr(R_ClassSymbol) = RStrings::get().dataframe_str;
+		dest_list.attr(R_RowNamesSymbol) = {NA_INTEGER, -static_cast<int>(nrows)};
```
Is there a more idiomatic way to return a data frame with cpp11?

james-atkins replied:
This is intentional - will add a comment to explain. cpp11's [data frame constructor](https://github.com/r-lib/cpp11/blob/322e5ee67346926993cc6f464152802118250b78/inst/include/cpp11/data_frame.hpp#L70) wouldn't work here as it tries to calculate the number of rows itself, and gives the wrong answer if the first column is itself another data frame or if the struct is empty.

krlmlr commented on tools/rpkg/R/dbConnect__duckdb_driver.R:
```diff
@@ -9,6 +9,9 @@
 #' @param read_only Set to `TRUE` for read-only operation
+#' @param tibble Whether a [tibble::tbl_df] should be returned rather than
+#'   a [data.frame]. This is recommended if you are working with lists and structs
+#'   as tibbles, unlike regular data frames, can easily handle nested data.
```
Minor: nested columns are lists of data frames (or other vectors), structs would correspond to packed columns.

```suggestion
#'   as tibbles, unlike regular data frames, can easily handle nested and packed data.
```

krlmlr reviewed (commented):
Thanks for taking this on!

rsund commented:
Thanks for your work on this certainly useful feature. I have a few comments for your consideration. Sorry that some of these are just (non tested) ideas, but most likely it would be valuable to think these through:

Probably there is no need to add `tibble` for the dependencies, because it is also exported from `dplyr` that is already among the suggested packages. See the discussion related to the use of `tibble` in #3276 and also check [how to use function pkg_method to use functions from suggested packages so that no R NOTEs are introduced](https://github.com/duckdb/duckdb/pull/3318/commits/87cf0e3a128f34ac64a999f7fa5948ea8a7e862b).  More examples in [backend-dbplyr__duckdb_connection.R](https://github.com/duckdb/duckdb/blob/master/tools/rpkg/R/backend-dbplyr__duckdb_connection.R).

I also think that this could be doable with `data.frame` alone that would be a preferable option in the case that structs need to be transferred into R with only the minimal imported dependencies. If "nested" data.frames are not working nicely, how about using just lists or maybe adding a customized print method if that was the issue?

And what happens if you try to use the exported `tibble` in a vanilla environment? E.g. what happens when `tibble` is used only with `data.frame` class (each `tibble` is also a `data.frame`)? Maybe you could just generate the similar `data.frame` in DuckDB without using tools from the `tibble`-package (i.e. without additional dependencies), but still adding `tibble`-class manually to the generated `data.frame` so then it would be possible to generate wanted structures in DuckDB without dependencies, but the benefits of those dependencies would still be automatically available "outside" of DuckDB to display these special `data.frames` nicely? **Please consider this potential option carefully, because this could maybe be the best option to "get it all" without introducing any new dependencies.**

And you mentioned that `data.frame` would be the most intuitive option to export structs. What would be the other non-intuitive options? Certainly `lists` but are there some other options?

What if there are nested structs in DuckDB - are those now exported correctly? Is the exporting of structs now "compatible" with exporting of DuckDB lists?

Have you tested would it be possible to use `reticulate`-package to export structs via DuckDB Python client to R? That is certainly not the solution here, but if it would be possible to get structs to R that way, then maybe it could offer some insights to suitable R data structures to be used.

hannes commented:
I would also prefer a data.frame transformation by default

krlmlr commented:
Construction of nested and packed data frames in R is painful or impossible. In the example below I use `vctrs::data_frame()` to avoid that pain. Once they are constructed, basic subsetting operations seem to work, though. I agree that we can use plain data frames as a container in the default case. 

``` r
plain <- vctrs::data_frame(x = 0, a = 1:3)
nested_list <- vctrs::data_frame(x = 0, a = list(1:3, 4:6, 7:9))
nested_df <- vctrs::data_frame(x = 0, a = list(vctrs::data_frame(b = 1:3, c = 3:1), vctrs::data_frame(b = 4:6, c = 6:4)))
packed_df <- vctrs::data_frame(x = 0, a = vctrs::data_frame(b = 1:3, c = 4:6))

plain
#>   x a
#> 1 0 1
#> 2 0 2
#> 3 0 3
nested_list
#>   x       a
#> 1 0 1, 2, 3
#> 2 0 4, 5, 6
#> 3 0 7, 8, 9
nested_df
#>   x                a
#> 1 0 1, 2, 3, 3, 2, 1
#> 2 0 4, 5, 6, 6, 5, 4
packed_df
#>   x a.b a.c
#> 1 0   1   4
#> 2 0   2   5
#> 3 0   3   6

waldo::compare(plain[1, , drop = FALSE], vctrs::vec_slice(plain, 1))
#> ✓ No differences
waldo::compare(nested_list[1, , drop = FALSE], vctrs::vec_slice(nested_list, 1))
#> ✓ No differences
waldo::compare(nested_df[1, , drop = FALSE], vctrs::vec_slice(nested_df, 1))
#> ✓ No differences
waldo::compare(packed_df[1, , drop = FALSE], vctrs::vec_slice(packed_df, 1))
#> ✓ No differences

waldo::compare(plain[, 2, drop = FALSE], vctrs::data_frame(a = plain[[2]]))
#> ✓ No differences
waldo::compare(nested_list[, 2, drop = FALSE], vctrs::data_frame(a = nested_list[[2]]))
#> ✓ No differences
waldo::compare(nested_df[, 2, drop = FALSE], vctrs::data_frame(a = nested_df[[2]]))
#> ✓ No differences
waldo::compare(packed_df[, 2, drop = FALSE], vctrs::data_frame(a = packed_df[[2]]))
#> ✓ No differences
```

<sup>Created on 2022-05-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

krlmlr commented:
Converting deeply nested data frames to deeply nested tibbles is painful, this is something duckdb or even DBI might want to offer optionally, perhaps through a connection option -- as in this PR. We want to ensure that this connection option only works if the tibble package can be loaded, to ensure that all tibble methods are available in the session.

Not for this PR: the nested data frame (list column with data frames in each list element) has a memory overhead of 1-2 orders of magnitude, and is painfully slow to construct. I wonder if we could make nesting a lazy operation that returns an ALTREP list where the interior data frames are materialized only on demand, so that perhaps nest + unnest works without ever materializing interior data frames.

hannes commented:
there are no ALTREP lists regrettably

krlmlr commented:
...yet?

krlmlr commented:
I think we can construct proper nested and packed data frames in C++ code without adding new dependencies. I used `vctrs::data_frame()` because constructing these objects via `data.frame()` is awkward or impossible.

krlmlr commented:
Taking a closer look here now. Looks good overall, still fighting a corner case.

james-atkins commented:
I have some (unfinished and unpushed) changes where I set the correct attributes on the C++ side and it seems to work with plain data frames, at least in simple cases - the culprit seemed to be the `as.data.frame` call on the R side but we can dispense with that. Interested to see your corner case.

I have a particularly nasty real-world parquet file with about 3 layers of structs inside structs and lists of structs that I will try and use for some tests.

krlmlr commented:
Thanks! I have pushed to https://github.com/krlmlr/duckdb/tree/r_struct, removing the tibble-related changes, tweaking tests and fixing one corner case. Would you like to cherry-pick from there?

I'll add tests that check that the behavior via Arrow is identical to the native data frame behavior.

krlmlr commented:
Looks like `list_to_df()` is still called somewhere?

I submitted https://github.com/r-lib/cpp11/pull/272 so that we could use cpp11's `data_frame()` at some point.

krlmlr reviewed (approved):
Looks great, thanks!

krlmlr commented:
I wonder why the R tests fail, though. Can you please merge the main branch and push to trigger the tests again?

james-atkins committed:
R: support structs in queries

This PR allows structs to be returned when running queries from R.
DuckD structs are converted to R data.frames that are nested or packed
inside the primary data frame returned from DuckDB.

As the base R as.data.frame method does not handle nested or packed data
frames well and as DuckDB guarantees that the output will always be a
valid data frame, the necessary attributes (class and number of
rows) are set from C++.

krlmlr reviewed (approved):
Tests look good now, thanks!

hannes commented on tools/rpkg/tests/testthat/test_struct.R:
```diff
@@ -0,0 +1,110 @@
+  on.exit(dbDisconnect(con, shutdown = TRUE))
+
+  res <- dbGetQuery(con, "SELECT {'x': 100, 'y': 'hello', 'z': 3.14} AS s")
+  expect_equal(res, vctrs::data_frame(
```
Here and below, the tests use vctrs and tibble, are both required or could one of them (or base R even) suffice?

krlmlr replied:
Constructing nested and packed data frames is ugly in base R, it requires the use of `I()` which will break the subsequent comparison and also creates non-integer row names in one case:

``` r
dput(data.frame(a = I(data.frame(b = 1, c = 2))))
#> structure(list(a = structure(list(b = 1, c = 2), class = c("AsIs", 
#> "data.frame"), row.names = c(NA, -1L))), class = "data.frame", row.names = "1")
dput(data.frame(a = I(list(data.frame(b = 1:2, c = 2:3)))))
#> structure(list(a = structure(list(structure(list(b = 1:2, c = 2:3), class = "data.frame", row.names = c(NA, 
#> -2L))), class = "AsIs")), class = "data.frame", row.names = c(NA, 
#> -1L))
```

<sup>Created on 2022-06-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

If we don't want to bring in vctrs, we can hand-roll a constructor that doesn't have this limitation, but read on.

The arrow package returns tibbles and "list_of" vectors. Implementing the comparison without the tibble package (which, like arrow, imports vctrs), will require substantial workarounds.

hannes replied:
Ok thanks for the explanation

hannes commented:
LGTM, could you please resolve the merge conflict?

krlmlr commented:
`git merge origin/master -X ours` should do it, the conflict is because we moved `list_to_df()` to a different place and it has received a no-op change from elsewhere.

james-atkins commented:
Thanks, fixed the merge conflict

krlmlr commented:
I'm starting to work on `nest()`, `unnest()`, `pack()` and `unpack()` based on this. Can't wait to see it merged!

krlmlr commented:
Prototype `nest()` and `pack()`, works beautifully: https://github.com/duckdb/duckdb/commit/bf1c5a8be5d6cf9b87be521dc709c9f3a8371c4d.

`unnest()` and `unpack()` will be a bit trickier, we will need to run the query with a `LIMIT 0` to get an idea of the data structure.

Thanks again @james-atkins for your work on this!

krlmlr commented:
FWIW, the implementation currently creates a subquery for `nest()`. Should we inline instead?

``` r
duck <- DBI::dbConnect(duckdb::duckdb())

tbl <- dplyr::copy_to(duck, palmerpenguins::penguins, "penguins")

tbl |>
  tidyr::nest(data = -species) |>
  dbplyr::sql_render()
#> <SQL> SELECT "species", LIST("data") AS "data"
#> FROM (
#>   SELECT
#>     "species",
#>     ROW("island", "bill_length_mm", "bill_depth_mm", "flipper_length_mm", "body_mass_g", "sex", "year") AS "data"
#>   FROM "penguins"
#> ) "q01"
#> GROUP BY "species"
```

<sup>Created on 2022-06-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

rsund commented:
> Prototype `nest()` and `pack()`, works beautifully: [krlmlr@bf1c5a8](https://github.com/krlmlr/duckdb/commit/bf1c5a8be5d6cf9b87be521dc709c9f3a8371c4d).
> 
> `unnest()` and `unpack()` will be a bit trickier, we will need to run the query with a `LIMIT 0` to get an idea of the data structure.

Nice! Hopefully the DuckDB's own super-useful [unnest](https://duckdb.org/docs/sql/query_syntax/unnest) won't cause any namespace conflicts or masking here as at least I use it (DuckDB's own unnest) a lot in my R+dbplyr workflows.
